internal/services/containerapps/container_app_resource.go

```diff
@@ -0,0 +1,414 @@
+				}
+			}
+
+			state.Secrets = helpers.FlattenContainerAppSecrets(secretsResp.Model)
+
+			return metadata.Encode(&state)
+		},
+	}
+}
+
+func (r ContainerAppResource) Delete() sdk.ResourceFunc {
+	return sdk.ResourceFunc{
+		Timeout: 30 * time.Minute,
+		Func: func(ctx context.Context, metadata sdk.ResourceMetaData) error {
+			client := metadata.Client.ContainerApps.ContainerAppClient
+
+			id, err := containerapps.ParseContainerAppID(metadata.ResourceData.Id())
+			if err != nil {
+				return err
+			}
+
+			resp, err := client.Delete(ctx, *id)
+			if err != nil {
+				if !response.WasNotFound(resp.HttpResponse) {
+					return fmt.Errorf("deleting %s: %+v", *id, err)
+				}
+			}
+
+			if err := resp.Poller.PollUntilDone(); err != nil {
+				return fmt.Errorf("waiting for deletion of %s", *id)
+			}
+
+			return nil
+		},
+	}
+}
+
+func (r ContainerAppResource) Update() sdk.ResourceFunc {
+	return sdk.ResourceFunc{
+		Timeout: 30 * time.Minute,
+		Func: func(ctx context.Context, metadata sdk.ResourceMetaData) error {
+			client := metadata.Client.ContainerApps.ContainerAppClient
+
+			id, err := containerapps.ParseContainerAppID(metadata.ResourceData.Id())
+			if err != nil {
+				return err
+			}
+
+			var state ContainerAppModel
+			if err := metadata.Decode(&state); err != nil {
+				return err
+			}
+
+			existing, err := client.Get(ctx, *id)
+			if err != nil || existing.Model == nil {
+				return fmt.Errorf("reading %s: %+v", *id, err)
+			}
+
+			model := existing.Model
+
+			// Delta-updates need the secrets back from the list API, or we'll end up removing them or erroring out.
+			secretsResp, err := client.ListSecrets(ctx, *id)
+			if err != nil || secretsResp.Model == nil {
+				if secretsResp.HttpResponse == nil || secretsResp.HttpResponse.StatusCode != http.StatusNoContent {
+					return fmt.Errorf("retrieving secrets for update for %s: %+v", *id, err)
+				}
+			}
+			model.Properties.Configuration.Secrets = helpers.UnpackContainerSecretsCollection(secretsResp.Model)
+
+			if metadata.ResourceData.HasChange("revision_mode") {
+				revisionMode := containerapps.ActiveRevisionsMode(state.RevisionMode)
+				model.Properties.Configuration.ActiveRevisionsMode = &revisionMode
+			}
+
+			if metadata.ResourceData.HasChange("ingress") {
+				model.Properties.Configuration.Ingress = helpers.ExpandContainerAppIngress(state.Ingress, id.ContainerAppName)
+			}
+
+			if metadata.ResourceData.HasChange("registry") {
+				model.Properties.Configuration.Registries = helpers.ExpandContainerAppRegistries(state.Registries)
+
+			}
+
+			if metadata.ResourceData.HasChange("dapr") {
+				model.Properties.Configuration.Dapr = helpers.ExpandContainerAppDapr(state.Dapr)
+
+			}
+
+			if metadata.ResourceData.HasChange("template") {
+				allProbesRemoved := helpers.ContainerAppProbesRemoved(metadata)
+				if allProbesRemoved {
+					nilProbes := make([]containerapps.ContainerAppProbe, 0)
+					containers := *model.Properties.Template.Containers
```

Comment:
there's a crash here if `model.Properties.Template` or `model.Properties.Template.Containers` is nil/not returned?